In [2]:
import pandas as pd
import yfinance as yf
import numpy as np
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [3]:
data = yf.download("AAPL",start='2025-01-01')
data.columns = data.columns.get_level_values(0)  # flatten to single level
data

[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Date,,,,,
2025-01-02,242.301926,247.518596,240.284814,247.349662,55740700
2025-01-03,241.815033,242.629819,240.354364,241.815033,40244100
2025-01-06,243.444611,245.759820,241.656035,242.758989,45045600
2025-01-07,240.672318,243.991110,239.817777,241.437418,40856000
2025-01-08,241.159210,242.162808,238.526040,240.384163,37628900
...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800


In [4]:
data['Returns'] = data['Close'].pct_change()
data = data.dropna()
data


Price,Close,High,Low,Open,Volume,Returns
Date,,,,,,
2025-01-03,241.815033,242.629819,240.354364,241.815033,40244100,-0.002009
2025-01-06,243.444611,245.759820,241.656035,242.758989,45045600,0.006739
2025-01-07,240.672318,243.991110,239.817777,241.437418,40856000,-0.011388
2025-01-08,241.159210,242.162808,238.526040,240.384163,37628900,0.002023
2025-01-10,235.346359,238.635343,231.520795,238.486286,61710900,-0.024104
...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557


# **SMA ANALYSIS**

In [5]:
data['SMA_20'] = data['Close'].rolling(20).mean()
data['SMA_50'] = data['Close'].rolling(50).mean()


/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_39119/2929945010.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['SMA_20'] = data['Close'].rolling(20).mean()
/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_39119/2929945010.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['SMA_50'] = data['Close'].rolling(50).mean()


In [6]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-01-03,241.815033,242.629819,240.354364,241.815033,40244100,-0.002009,NaN,NaN
2025-01-06,243.444611,245.759820,241.656035,242.758989,45045600,0.006739,NaN,NaN
2025-01-07,240.672318,243.991110,239.817777,241.437418,40856000,-0.011388,NaN,NaN
2025-01-08,241.159210,242.162808,238.526040,240.384163,37628900,0.002023,NaN,NaN
2025-01-10,235.346359,238.635343,231.520795,238.486286,61710900,-0.024104,NaN,NaN
...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.9922
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.8296
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.5888


In [7]:
data.dropna(inplace=True)

/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_39119/1368182302.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.dropna(inplace=True)


In [8]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-03-18,211.572113,214.019175,210.378423,213.034388,42432400,-0.006121,231.235704,232.228108
2025-03-19,214.108704,217.610192,212.626530,213.094060,54385400,0.011989,229.761992,231.673982
2025-03-20,212.974701,216.346883,211.104577,212.865278,48862900,-0.005296,228.183831,231.064583
2025-03-21,217.122772,217.689768,210.169506,210.448033,94127800,0.019477,226.827000,230.593593
2025-03-24,219.569839,220.315897,217.431146,219.838425,44299500,0.011270,225.515430,230.161805
...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800


In [9]:
fig = make_subplots(rows=2,cols=1)
fig.add_trace(go.Candlestick(x=data.index,open=data['Open'],high=data['High'],low=data['Low'],close=data['Close'],name='Closing Price'),row=1,col=1)
fig.add_trace(go.Scatter(x=data.index,y=data['SMA_20'],name='SMA_20',mode='lines'),row=1,col=1)
fig.add_trace(go.Scatter(x=data.index,y=data['SMA_50'],name='SMA_50',mode='lines'),row=1,col=1)
fig.update_layout(
    height = 600,
    width=800,
    title='AAPL Price with Moving Averages',
    xaxis_title='Date',
    yaxis_title='Price (USD)',
    template='plotly_dark'  # looks clean
)
fig.add_trace(go.Bar(x=data.index,y=data['Volume'],name='Volume'),row=2,col=1)
fig.show()

# **TENDENCIES NOTED FROM THE CHART:**
1)The price tends to go up when the SMA20 marks overtakes SMA50.This was noticed on **17 July 2025** when the overall trend was upwards all the way until **6 January 2026** when the SMA divergence finished

2)The price tends to 'bounce' off the SMA50 line which proves the hypothesis of SMA having an impact on the price action

In [10]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-03-18,211.572113,214.019175,210.378423,213.034388,42432400,-0.006121,231.235704,232.228108
2025-03-19,214.108704,217.610192,212.626530,213.094060,54385400,0.011989,229.761992,231.673982
2025-03-20,212.974701,216.346883,211.104577,212.865278,48862900,-0.005296,228.183831,231.064583
2025-03-21,217.122772,217.689768,210.169506,210.448033,94127800,0.019477,226.827000,230.593593
2025-03-24,219.569839,220.315897,217.431146,219.838425,44299500,0.011270,225.515430,230.161805
...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800


# **VOLATILITY ANALYSIS**


In [11]:
data['Volatility'] = data['Returns'].rolling(20).std() * (252 ** 0.5)
data.dropna(inplace=True)
data

/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_39119/4174629229.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Volatility'] = data['Returns'].rolling(20).std() * (252 ** 0.5)
/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_39119/4174629229.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.dropna(inplace=True)


Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility
Date,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618185,0.792022
2025-04-15,201.077530,202.440324,198.749833,200.799003,51343900,-0.001876,207.011697,222.108486,0.791844
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968708,221.347023,0.801010
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117207,220.646034,0.803486
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868306,219.854141,0.800923
...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877


In [12]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=data.index, y=data['Close'],
              name='Close', mode='lines'))

fig.add_trace(go.Scatter(x=data.index, y=data['Volatility'],
              name='Volatility (20d)', mode='lines', yaxis='y2'))

fig.update_layout(
    title='AAPL Price & Volatility',
    yaxis2=dict(overlaying='y', side='right'),  # second y-axis
    template='plotly_dark'
)
fig.show()


# **TENDENCIES SPOTTED**

1)The price tends to go up after a drop in volatility

2)The trend becomes stronger when volatility consolidates

3)Volatility spikes usually happen after a big price drop

In [13]:
volat_testdata = data.copy()
volat_testdata['Vol_MA'] = volat_testdata['Volatility'].rolling(20).mean()
volat_testdata['Vol_drop'] = volat_testdata['Volatility'] < volat_testdata['Vol_MA']
volat_testdata['Price after 5d'] = volat_testdata['Returns'].shift(-5)
volat_testdata.dropna(inplace=True)
volat_testdata

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Vol_MA,Vol_drop,Price after 5d
Date,,,,,,,,,,,,
2025-05-12,209.957047,210.435161,205.933017,210.136343,63775800,0.063146,202.147723,209.314560,0.384036,0.720323,True,-0.011739
2025-05-13,212.088593,212.556737,208.174129,209.598471,51909300,0.010152,202.679375,208.820754,0.378671,0.699656,True,-0.009196
2025-05-14,211.490967,213.094605,209.747882,211.590563,49325800,-0.002818,203.200047,208.356775,0.378841,0.679006,True,-0.023059
2025-05-15,210.614456,212.118499,208.712000,210.116432,45029500,-0.004144,204.068324,207.879045,0.346787,0.656294,True,-0.003612
2025-05-16,210.425201,211.730038,208.941099,211.520861,54737900,-0.000899,204.792350,207.405686,0.345443,0.633392,True,-0.030244
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-20,326.589996,333.709991,323.679993,333.510010,53468000,-0.021424,306.946999,303.419308,0.375896,0.341221,False,0.011681
2026-07-21,327.739990,329.600006,322.220001,323.130005,41338900,0.003521,308.483498,304.230600,0.374694,0.347680,False,0.009409
2026-07-22,325.890015,329.000000,323.339996,327.869995,38755900,-0.005645,310.062999,304.887400,0.373130,0.354266,False,-0.005557


In [14]:
volat_testdata.groupby('Vol_drop')['Price after 5d'].mean()

Vol_drop
False    0.002282
True     0.000744
Name: Price after 5d, dtype: float64

In [15]:
volat_testdata

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Vol_MA,Vol_drop,Price after 5d
Date,,,,,,,,,,,,
2025-05-12,209.957047,210.435161,205.933017,210.136343,63775800,0.063146,202.147723,209.314560,0.384036,0.720323,True,-0.011739
2025-05-13,212.088593,212.556737,208.174129,209.598471,51909300,0.010152,202.679375,208.820754,0.378671,0.699656,True,-0.009196
2025-05-14,211.490967,213.094605,209.747882,211.590563,49325800,-0.002818,203.200047,208.356775,0.378841,0.679006,True,-0.023059
2025-05-15,210.614456,212.118499,208.712000,210.116432,45029500,-0.004144,204.068324,207.879045,0.346787,0.656294,True,-0.003612
2025-05-16,210.425201,211.730038,208.941099,211.520861,54737900,-0.000899,204.792350,207.405686,0.345443,0.633392,True,-0.030244
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-20,326.589996,333.709991,323.679993,333.510010,53468000,-0.021424,306.946999,303.419308,0.375896,0.341221,False,0.011681
2026-07-21,327.739990,329.600006,322.220001,323.130005,41338900,0.003521,308.483498,304.230600,0.374694,0.347680,False,0.009409
2026-07-22,325.890015,329.000000,323.339996,327.869995,38755900,-0.005645,310.062999,304.887400,0.373130,0.354266,False,-0.005557


In [16]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility
Date,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618185,0.792022
2025-04-15,201.077530,202.440324,198.749833,200.799003,51343900,-0.001876,207.011697,222.108486,0.791844
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968708,221.347023,0.801010
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117207,220.646034,0.803486
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868306,219.854141,0.800923
...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877


In [17]:
sma_strategy = data.copy()


In [18]:
sma_strategy['Signal'] = np.where(sma_strategy['SMA_20'] > sma_strategy['SMA_50'],1,0)
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal
Date,,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618185,0.792022,0
2025-04-15,201.077530,202.440324,198.749833,200.799003,51343900,-0.001876,207.011697,222.108486,0.791844,0
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968708,221.347023,0.801010,0
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117207,220.646034,0.803486,0
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868306,219.854141,0.800923,0
...,...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070,1
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872,1
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877,1


In [19]:
sma_strategy['Strategy_Returns'] = sma_strategy['Signal'].shift(1) * sma_strategy['Returns']
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal,Strategy_Returns
Date,,,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618185,0.792022,0,NaN
2025-04-15,201.077530,202.440324,198.749833,200.799003,51343900,-0.001876,207.011697,222.108486,0.791844,0,-0.000000
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968708,221.347023,0.801010,0,-0.000000
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117207,220.646034,0.803486,0,0.000000
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868306,219.854141,0.800923,0,-0.000000
...,...,...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070,1,0.011681
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872,1,0.009409
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877,1,-0.005557


In [20]:
sma_strategy.groupby('Signal')['Strategy_Returns'].mean()

Signal
0   -0.000276
1    0.001322
Name: Strategy_Returns, dtype: float64

In [21]:
sharpe = (sma_strategy['Strategy_Returns'].mean() /
          sma_strategy['Strategy_Returns'].std() * (252 ** 0.5))
print(f'Sharpe: {sharpe:.2f}')

Sharpe: 0.96


In [22]:
sma_strategy['Equity_Curve'] = (1 + sma_strategy['Strategy_Returns']).cumprod()
sma_strategy['Buy_Hold'] = (1 + sma_strategy['Returns']).cumprod()
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal,Strategy_Returns,Equity_Curve,Buy_Hold
Date,,,,,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618185,0.792022,0,NaN,NaN,1.022054
2025-04-15,201.077530,202.440324,198.749833,200.799003,51343900,-0.001876,207.011697,222.108486,0.791844,0,-0.000000,1.000000,1.020136
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968708,221.347023,0.801010,0,-0.000000,1.000000,0.980419
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117207,220.646034,0.803486,0,0.000000,1.000000,0.994095
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868306,219.854141,0.800923,0,-0.000000,1.000000,0.974817
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070,1,0.011681,1.382648,1.709261
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872,1,0.009409,1.395657,1.725344
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877,1,-0.005557,1.387901,1.715755


In [23]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=sma_strategy.index,y=sma_strategy['Equity_Curve'],name='Equity Curve',mode='lines'))
fig.add_trace(go.Scatter(x=sma_strategy.index,y=sma_strategy['Buy_Hold'],name='Buy & Hold',mode='lines'))
fig.update_layout(title='Strategy vs Buy & Hold',
                  template='plotly_dark')
fig.show()

In [24]:
bh_sharpe = sma_strategy['Returns'].mean() / (sma_strategy['Returns'].std() * (252 ** 0.5))
bh_sharpe

np.float64(0.00575813930578003)

In [31]:
sma_quick = data.copy()
sma_quick.drop(['SMA_20','SMA_50'],axis=1,inplace=True)
sma_quick

Price,Close,High,Low,Open,Volume,Returns,Volatility
Date,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,0.792022
2025-04-15,201.077530,202.440324,198.749833,200.799003,51343900,-0.001876,0.791844
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,0.801010
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,0.803486
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,0.800923
...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,0.290070
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,0.283872
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,0.280877


In [35]:
sma_quick['SMA_10'] = sma_quick['Close'].rolling(10).mean()
sma_quick['SMA_30'] = sma_quick['Close'].rolling(30).mean()

sma_quick['Signal'] = np.where(sma_quick['SMA_10'] > sma_quick['SMA_30'], 1, 0)
sma_quick.dropna(inplace=True)
sma_quick['Signal_Returns'] = sma_quick['Signal'].shift(1) * sma_quick['Returns']
sma_quick.dropna(inplace=True)

In [36]:
sma_quick.groupby('Signal')['Signal_Returns'].mean()

Signal
0    0.000129
1    0.001385
Name: Signal_Returns, dtype: float64

In [42]:
def sharpe(returns):
    return returns.mean() / returns.std() * (252 ** 0.5)



In [39]:
sma_quick['Buy_Hold'] = (1 + sma_quick['Returns']).cumprod()
sma_quick['Equity_Curve'] = (1 + sma_quick['Signal_Returns']).cumprod()
sma_quick

Price,Close,High,Low,Open,Volume,Returns,Volatility,SMA_10,SMA_30,Signal,Signal_Returns,Buy_Hold,Equity_Curve
Date,,,,,,,,,,,,,
2025-08-20,225.372391,229.819815,225.133078,229.331192,42263900,-0.019735,0.306605,228.439781,215.765031,1,-0.019735,0.980265,0.980265
2025-08-21,224.265518,225.880958,223.148683,225.631664,30621200,-0.004911,0.307435,228.950279,216.188194,1,-0.004911,0.975451,0.975451
2025-08-22,227.117462,228.443712,224.774101,225.531951,42477800,0.012717,0.309296,228.817653,216.747923,1,0.012717,0.987856,0.987856
2025-08-25,226.519165,228.653127,225.591781,225.841075,30983100,-0.002634,0.309915,228.815660,217.372041,1,-0.002634,0.985253,0.985253
2025-08-26,228.663086,228.842586,224.056124,226.229967,54575100,0.009465,0.304551,228.781757,218.051353,1,0.009465,0.994578,0.994578
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,0.290070,328.116998,308.830333,1,0.011681,1.465402,1.369357
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,0.283872,330.638998,310.462000,1,0.009409,1.479190,1.382241
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,0.280877,331.707999,311.854332,1,-0.005557,1.470970,1.374559


In [40]:
fig = go.Figure()

fig.add_trace(go.Scatter(x = sma_quick.index,y=sma_quick['Equity_Curve'],mode='lines',name='Equity Curve'))
fig.add_trace(go.Scatter(x = sma_quick.index,y=sma_quick['Buy_Hold'],mode='lines',name='Buy_Hold'))

fig.update_layout(title='Strategy vs Buy & Hold',
                  template='plotly_dark')

fig.show()

In [43]:
print(f'Sharpe ratio of strategy: {sharpe(sma_quick['Signal_Returns'])}')
print(f'Sharpe ratio of holding: {sharpe(sma_quick['Returns'])}')

Sharpe ratio of strategy: 1.1889253199877405
Sharpe ratio of holding: 1.3176671079048565


## Comparison: SMA 20/50 vs SMA 10/30 vs Buy & Hold

| Strategy   | Sharpe |
|------------|--------|
| SMA 20/50  | 1.30   |
| SMA 10/30  | 1.19   |
| Buy & Hold | 1.32   |

Conclusion: On a strongly trending stock like AAPL (2025–2026),
buy and hold outperforms both SMA crossover strategies on a 
risk-adjusted basis. SMA crossover strategies are better suited
to sideways or choppy markets.

Next step: test on a sideways/choppy stock to verify this hypothesis.